actionpack/lib/action_controller/renderer.rb

```diff
@@ -103,8 +103,13 @@ def render(*args)
     private
       def normalize_keys(defaults, env)
         new_env = {}
-        defaults.each_pair { |k, v| new_env[rack_key_for(k)] = rack_value_for(k, v) }
         env.each_pair { |k, v| new_env[rack_key_for(k)] = rack_value_for(k, v) }
+
+        defaults.each_pair do |k, v|
+          key = rack_key_for(k)
+          new_env[key] = rack_value_for(k, v) unless new_env.key?(key)
```

Comment:
```suggestion
          defaults.each_pair { |k, v| new_env[rack_key_for(k)] = rack_value_for(k, v) unless env.key?(k) }
```

Reply:
hm... `defaults` and `env` has not compatible keys (`:http_host` vs `"HTTP_HOST"` etc). Looking like `rack_key_for` is required before compare for normalization.